### ISSUE TITLE
The decoding result contains many < unk >

### ISSUE THREAD

**Issue by rookie0607** [wenet-e2e/wenet#936]
Hello, wenet team. I have encountered some problems in training with my own 1K data. Training without modifying any parameters of train.yaml, train_ Loss and CV_ Loss has been oscillating around 70, and the model cannot converge. So I keep trying to reduce the learning rate, but I encounter new problems. The decoding result contains many < unk >, how can I solve it?
The decoding results are as follows：
![image](https://user-images.githubusercontent.com/68495525/153257728-aa0be96b-b80d-49e8-9de7-c31473360c1a.png)

train.yaml as follows:
accum_grad: 1
cmvn_file: data/train_960/global_cmvn
dataset_conf:
  batch_conf:
    batch_size: 14
    batch_type: static
  fbank_conf:
    dither: 0.0
    frame_length: 25
    frame_shift: 10
    num_mel_bins: 80
  filter_conf:
    max_length: 2100
    max_output_input_ratio: 0.1
    min_length: 50
    min_output_input_ratio: 0.0005
    token_max_length: 400
    token_min_length: 1
  resample_conf:
    resample_rate: 16000
  shuffle: true
  shuffle_conf:
    shuffle_size: 1500
  sort: true
  sort_conf:
    sort_size: 500
  spec_aug: true
  spec_aug_conf:
    max_f: 10
    max_t: 50
    num_f_mask: 2
    num_t_mask: 2
  speed_perturb: true
decoder: transformer
decoder_conf:
  attention_heads: 4
  dropout_rate: 0.1
  linear_units: 2048
  num_blocks: 6
  positional_dropout_rate: 0.1
  self_attention_dropout_rate: 0.0
  src_attention_dropout_rate: 0.0
encoder: conformer
encoder_conf:
  activation_type: swish
  attention_dropout_rate: 0.0
  attention_heads: 4
  cnn_module_kernel: 15
  dropout_rate: 0.1
  input_layer: conv2d
  linear_units: 2048
  normalize_before: true
  num_blocks: 12
  output_size: 256
  pos_enc_layer_type: rel_pos
  positional_dropout_rate: 0.1
  selfattention_layer_type: rel_selfattn
  use_cnn_module: true
grad_clip: 5
input_dim: 80
is_json_cmvn: true
log_interval: 100
max_epoch: 100
model_conf:
  ctc_weight: 0.3
  length_normalized_loss: false
  lsm_weight: 0.1
optim: adam
optim_conf:
  lr: 8.0e-05
output_dim: 5048
scheduler: warmuplr
scheduler_conf:
  warmup_steps: 25000

The training log is as follows：
cv_loss: 74.36538192687848
epoch: 0
lr: 2.0e-09
step: 11447
cv_loss: 8.87004251305317
epoch: 10
lr: 3.413521024499168e-05
step: 148724
cv_loss: 8.171826801895412
epoch: 11
lr: 3.279955884890021e-05
step: 160159
cv_loss: 7.677792881704005
epoch: 12
lr: 3.16069770620507e-05
step: 171609
cv_loss: 7.365098930784381
epoch: 13
lr: 3.053435114503817e-05
step: 183034
cv_loss: 7.047641605855673
epoch: 14
lr: 2.956602362137194e-05
step: 194454
cv_loss: 6.703525782245951
epoch: 15
lr: 2.86847080496589e-05
step: 205864
cv_loss: 6.379399753996051
epoch: 16
lr: 2.7878457000074905e-05
step: 217318
cv_loss: 6.227978232697523
epoch: 17
lr: 2.7133832664137963e-05
step: 228755
cv_loss: 6.069486028959661
epoch: 18
lr: 2.6446837354797223e-05
step: 240184
cv_loss: 5.795553260737814
epoch: 19
lr: 2.5809943308642773e-05
step: 251610
cv_loss: 69.42665398107225
epoch: 1
lr: 2.2896000000000002e-05
step: 22870
cv_loss: 5.689312238259004
epoch: 20
lr: 2.521710225079886e-05
step: 263029
cv_loss: 5.4178879280816785
epoch: 21
lr: 2.4663650007186592e-05
step: 274452
cv_loss: 5.476332227320742
epoch: 22
lr: 2.4144932752356955e-05
step: 285899
cv_loss: 5.313708624957217
epoch: 23
lr: 2.3656631975966278e-05
step: 297329
cv_loss: 5.173749357376585
epoch: 24
lr: 2.3197470221838892e-05
step: 308738
cv_loss: 5.13750762583737
epoch: 25
lr: 2.2764821259771276e-05
step: 320197
cv_loss: 5.076591373240961
epoch: 26
lr: 2.2353765148346037e-05
step: 331624
cv_loss: 4.868604764188688
epoch: 27
lr: 2.1965260582253228e-05
step: 343035
cv_loss: 4.9403411222009455
epoch: 28
lr: 2.1596836554293287e-05
step: 354460
cv_loss: 4.782982207689635
epoch: 29
lr: 2.1245930777293284e-05
step: 365880
cv_loss: 68.98270676690416
epoch: 2
lr: 4.5742000000000004e-05
step: 34307
cv_loss: 4.735401319866874
epoch: 30
lr: 2.0911734801252404e-05
step: 377315
cv_loss: 4.704278237887265
epoch: 31
lr: 2.0592419667838934e-05
step: 388730
cv_loss: 4.556968025701016
epoch: 32
lr: 2.0287820959244413e-05
step: 400157
cv_loss: 4.619346503249422
epoch: 33
lr: 1.9996051169802448e-05
step: 411604
cv_loss: 4.650601996896193
epoch: 34
lr: 1.9716039086292663e-05
step: 423036
cv_loss: 4.464593855273683
epoch: 35
lr: 1.9447814997094783e-05
step: 434477
cv_loss: 4.470056243232299
epoch: 36
lr: 1.9190049545425624e-05
step: 445913
cv_loss: 4.460705865494574
epoch: 37
lr: 1.8942375323284573e-05
step: 457348
cv_loss: 4.397569977643154
epoch: 38
lr: 1.870407022585034e-05
step: 468806
cv_loss: 4.43527145077819
epoch: 39
lr: 1.8474085423818206e-05
step: 480214
cv_loss: 67.1390297268128
epoch: 3
lr: 4.2681758873262624e-05
step: 45751
cv_loss: 4.3597102277877
epoch: 40
lr: 1.825333105557128e-05
step: 491640
cv_loss: 4.330533396439864
epoch: 41
lr: 1.80399755521515e-05
step: 503074
cv_loss: 4.268325597167288
epoch: 42
lr: 1.7833788975433553e-05
step: 514507
cv_loss: 4.252162153010106
epoch: 43
lr: 1.763453146728734e-05
step: 525919
cv_loss: 4.279474965434304
epoch: 44
lr: 1.7442155257954564e-05
step: 537345
cv_loss: 4.227706478565146
epoch: 45
lr: 1.7255715871260877e-05
step: 548797
cv_loss: 4.224219410450871
epoch: 46
lr: 1.7074725531929182e-05
step: 560222
cv_loss: 4.188409822366765
epoch: 47
lr: 1.689972057440834e-05
step: 571641
cv_loss: 4.138784058144549
epoch: 48
lr: 1.673007648057971e-05
step: 583070
cv_loss: 4.113389953817289
epoch: 49
lr: 1.6565298668816778e-05
step: 594504
cv_loss: 65.70921286247577
epoch: 4
lr: 3.696025568211654e-05
step: 57173
cv_loss: 4.1118601765078235
epoch: 50
lr: 1.640522667654976e-05
step: 605923
cv_loss: 4.155135974526951
epoch: 51
lr: 1.6249908277143787e-05
step: 617348
cv_loss: 4.121046050922158
epoch: 52
lr: 1.6098841212400923e-05
step: 628748
cv_loss: 4.121110343654882
epoch: 53
lr: 1.5952227611312677e-05
step: 640171
cv_loss: 4.102042898006592
epoch: 54
lr: 1.580926407369392e-05
step: 651598
cv_loss: 4.016359465621318
epoch: 55
lr: 1.5670028515016785e-05
step: 663020
cv_loss: 4.037436826259137
epoch: 56
lr: 1.553446672859416e-05
step: 674445
cv_loss: 3.999622236446553
epoch: 57
lr: 1.5402329135340035e-05
step: 685872
cv_loss: 3.9993492261915202
epoch: 58
lr: 1.5273484831116756e-05
step: 697300
cv_loss: 4.008677669577855
epoch: 59
lr: 1.5147809937781872e-05
step: 708723
cv_loss: 64.32931183512405
epoch: 5
lr: 3.306288297992381e-05
step: 68604
cv_loss: 3.9376126638051434
epoch: 60
lr: 1.5025240126267871e-05
step: 720147
cv_loss: 3.9414827186456387
epoch: 61
lr: 1.490558796550883e-05
step: 731565
cv_loss: 3.948703644585978
epoch: 62
lr: 1.478881018360967e-05
step: 742981
cv_loss: 3.9748906618472772
epoch: 63
lr: 1.4674754522128966e-05
step: 754409
cv_loss: 3.9560958930170687
epoch: 64
lr: 1.4563181866911216e-05
step: 765859
cv_loss: 3.9896323983893565
epoch: 65
lr: 1.4453908384202902e-05
step: 777302
cv_loss: 3.962183607163981
epoch: 66
lr: 1.4347122927351664e-05
step: 788734
cv_loss: 3.9916788216206625
epoch: 67
lr: 1.4242769137576286e-05
step: 800158
cv_loss: 3.9322869093386714
epoch: 68
lr: 1.4140730458457323e-05
step: 811554
cv_loss: 3.84110643053246
epoch: 69
lr: 1.4041096121989254e-05
step: 822972
cv_loss: 14.022238879678994
epoch: 6
lr: 3.2000000000000005e-09
step: 103055
cv_loss: 12.132897142828671
epoch: 7
lr: 3.940246123340275e-05
step: 114471
cv_loss: 10.466929428902414
epoch: 8
lr: 3.738611658900696e-05
step: 125883
cv_loss: 9.872497601061205
epoch: 9
lr: 3.565124692528133e-05
step: 137313

**Comment by robin1001** [comment]
Apparently, the training  is not converged. Please check the data prepare, which is the common error. And you can compare your data prepare to the standard recipe like librispeech to figure out the problem.

**Comment by rookie0607** [comment]
> Apparently, the training is not converged. Please check the data prepare, which is the common error. And you can compare your data prepare to the standard recipe like librispeech to figure out the problem.

I'm sure the data is OK, and I've reduced my learning rate accordingly.

**Comment by BSen007** [comment]
Well! I am facing same kind of issues with Arabic common voice. With same data in ESPNET it is going good but with Wenet getting the same decoding issue. Used the librispeech train_conformer.yaml as config.

**Comment by rookie0607** [comment]
> Well! I am facing same kind of issues with Arabic common voice. With same data in ESPNET it is going good but with Wenet getting the same decoding issue. Used the librispeech train_conformer.yaml as config.

I'm sorry, my problem hasn't been solved until now. I can't give you guidance.

**Comment by BSen007** [comment]
Hi team, any progress/guidance on that?

**Comment by rookie0607** [comment]
> 嗨团队，有任何进展/指导吗？

My problem is solved, because my data transcription contains large and small English letters, while wenet converts small letters to uppercase letters by default. In the processor.py can be solved by modifying the BPE function.

**Comment by BSen007** [comment]
Thanks! well my transcriptions also contains same as using BW encoding for Arabic. example: wymkn >n tkwn AlEA}lAt mHAfZp >kvr mn AllAzm. However, I am using unigram not BPE. Do you think it is an issue here?

**Comment by rookie0607** [comment]
It should be. You can try it my way

**Comment by rookie0607** [comment]
> Thanks! well my transcriptions also contains same as using BW encoding for Arabic. example: wymkn >n tkwn AlEA}lAt mHAfZp >kvr mn AllAzm. However, I am using unigram not BPE. Do you think it is an issue here?

It should be. You can try it my way

**Comment by BSen007** [comment]
Can you please explain a bit what you change in processor.py? Thanks in advance

**Comment by rookie0607** [comment]
> chars = pattern.split(txt.upper())

chars = pattern.split(txt)
